### ISSUE TITLE
HFD5 links to incompatible Anaconda supplied hdf5.dll on Windows

### ISSUE THREAD

**Issue by krcools** [JuliaIO/HDF5.jl#323]
I was able to workaround this issue but thought it was worth reporting since I suspect many installations to be similar to mine.

I am running Julia on Windows together with the Anaconda Python installation. On installing, Anaconda adds a number of entries to the system path variable, among which `C:\Anaconda\Library\bin`. When the `HDF5.jl` dependency build script is run, it finds the version of `hfd5.dll` installed in this Anaconda directory and considers the dependency fulfilled.

There seems to be an inconsistency though between the Anaconda version and the version expected by `HDF5.jl` resulting in the package tests to fail.

As a workaround I temporarily removed the Anaconda subdir from my path and invoked `Pkg.build("HDF5")`. This downloaded the mingw compiled version and the subsequent testing round passed successfully.

Would it be possible to modify  `deps/build.jl` to skip the Anaconda supplied dll? Again, I imagine many people have both Anaconda and Julia on their machines.

**Comment by timholy** [comment]
> Would it be possible to modify deps/build.jl to skip the Anaconda supplied dll?

That sounds like a good idea. Since this issue comes down to paths, it's probably best done by someone who has your kind of installation---it would be rather difficult to test otherwise.

**Comment by krcools** [comment]
I'm no expert in the finesses of `BinDeps` but maybe something like this can help to sort out invalid copies:

```
M = zeros(UInt64,1)
m = zeros(UInt64,1)
p = zeros(UInt64,1)
r = ccall((:H5get_libversion, "hdf5"), Int, (Ptr{UInt64}, Ptr{UInt64}, Ptr{UInt64}), M,m,p)
v = VersionNumber("$(M[1]).$(m[1]).$(p[1])")
```

For the Anaconda version (after running `conda update --all`) this gives `v"1.8.15"` whereas for the WinRPM supplied version this gives `v"1.8.13"`. Must be a very subtle break of backwards compatibility... Or an ABI issue?

**Comment by simonster** [comment]
Probably an issue with the runtime. Julia is built with MinGW, whereas Anaconda libhdf5 may be built with VC++. There are [a couple places](https://github.com/JuliaIO/HDF5.jl/blob/668dd90c90f38ed83b11177e96aeb469e616f5e4/src/plain.jl#L1447-L1448) where we hand management of memory allocated by libhdf5 over to Julia. If Julia uses a different runtime from libhdf5, this will cause bad things to happen when that memory is freed. You can check if this is the problem by seeing if the crash goes away if you change `true` to `false` on those lines, although this will leak memory.

If my hunch here is correct, I see two options:

- Make sure we always use WinRPM-supplied `libhdf5` on Windows
- Copy the data in those places if we are on Windows, and then free the original buffer using `H5free_memory`

**Comment by tkelman** [comment]
It would be better to find some detectable sign of mingw vs msvc versions of hdf5 rather than going on library version. maybe removing the hdf5.dll alias would work short term.

This is yet another example of why various pieces of software permanently adding themselves to the windows path is always a bad idea.

**Comment by tkelman** [comment]
maybe a bad heuristic, but what about copying the data if the library name bindeps found doesn't start with a lib prefix?

**Comment by musm** [comment]
thanks @krcools for the workaround!!

**Comment by musm** [comment]
This is a really painful bug. It came up on gitter for a user, I had to link him to krcools workaround here.

**Comment by tkelman** [comment]
we should just revert #288.

**Comment by tkelman** [comment]
You should also stop letting random pieces of software permanently add themselves to your path and add incompatible dll's.

**Comment by musm** [comment]
You mean like random software like julia :P

**Comment by tkelman** [comment]
Julia doesn't add itself to your system-wide path. If you're adding Julia to your path, don't do that either.

**Comment by musm** [comment]
Fair enough, but then why does HD5F look for system DLLs in path. See my point?

**Comment by tkelman** [comment]
That's BinDeps default behavior. The non-lib-prefixed aliases were a bad idea and should never have been added since those are almost certainly not going to be compatible here.

**Comment by simonster** [comment]
Should have been fixed by #333. Comment if not.
